base/compiler/ssair/passes.jl

```diff
@@ -818,12 +821,19 @@ end
     applyTvar = applyT.var
     applyTbody = applyT.body
 
-    isa(applyTbody, DataType) || return nothing
+    arg = unwrap_unionall(arg)
+    applyTbody = unwrap_unionall(applyTbody)
+
+    (isa(arg, DataType) && isa(applyTbody, DataType)) || return nothing
+
     applyTbody.name == arg.name || return nothing
-    length(applyTbody.parameters) == length(arg.parameters) == 1 || return nothing
-    applyTbody.parameters[1] === applyTvar || return nothing
-    arg.parameters[1] === tvar || return nothing
-    return LiftedValue(argdef.args[3])
+    length(applyTbody.parameters) == length(arg.parameters) || return nothing
+    for i = 1:length(applyTbody.parameters)
+        if applyTbody.parameters[i] === applyTvar && arg.parameters[i] === tvar
+            return LiftedValue(argdef.args[3])
+        end
+    end
```

Comment:
I wonder if it it may be clearer (and more extensible later) if written this way:
```suggestion
    # find a possible parameter number we are extracting when matching with `tvar`
    i = findfirst(===(tvar), arg.parameters)
    # determine if we can find the apply_type for that parameter
    if applyTvar === applyTbody.parameters[i]
        return LiftedValue(argdef.args[3])
    end
```

Reply:
I'm not sure whether we want to look at all the parameters or just the first one. We can probably look at the first one in applyTbody, but I think `arg` might potentially have multiple instances.

Reply:
First is indeed probably the most common to extract